module/zfs/spa.c

```diff
@@ -3210,6 +3210,62 @@ spa_add_spares(spa_t *spa, nvlist_t *config)
 	}
 }
 
+/* Update a nvlist with latest extended stats values. */
+static void
+spa_update_l2cache_stats_ex(vdev_t *vd, nvlist_t *l2cache) {
+
+	vdev_stat_ex_t *vsx;
+	int i;
+	uint64_t *ptr;
+	uint_t len;
+	struct {
+		const char *name;
+		unsigned int offset;
+	} list[] = {
+		{ZPOOL_CONFIG_QUEUE_ACTIVE,
+		    offsetof(vdev_stat_ex_t, vsx_queue_active)},
+		{ZPOOL_CONFIG_TOT_R_LAT_HISTO, offsetof(vdev_stat_ex_t,
+		    vsx_total_histo[ZIO_TYPE_READ])},
+		{ZPOOL_CONFIG_TOT_W_LAT_HISTO, offsetof(vdev_stat_ex_t,
+		    vsx_total_histo[ZIO_TYPE_WRITE])},
+		{ZPOOL_CONFIG_DISK_R_LAT_HISTO, offsetof(vdev_stat_ex_t,
+		    vsx_disk_histo[ZIO_TYPE_READ])},
+		{ZPOOL_CONFIG_DISK_W_LAT_HISTO, offsetof(vdev_stat_ex_t,
+		    vsx_disk_histo[ZIO_TYPE_WRITE])},
+		{ZPOOL_CONFIG_SYNC_R_LAT_HISTO, offsetof(vdev_stat_ex_t,
+		    vsx_queue_histo[ZIO_PRIORITY_SYNC_READ])},
+		{ZPOOL_CONFIG_SYNC_W_LAT_HISTO, offsetof(vdev_stat_ex_t,
+		    vsx_queue_histo[ZIO_PRIORITY_SYNC_WRITE])},
+		{ZPOOL_CONFIG_ASYNC_R_LAT_HISTO, offsetof(vdev_stat_ex_t,
+		    vsx_queue_histo[ZIO_PRIORITY_ASYNC_READ])},
+		{ZPOOL_CONFIG_ASYNC_W_LAT_HISTO, offsetof(vdev_stat_ex_t,
+		    vsx_queue_histo[ZIO_PRIORITY_ASYNC_WRITE])},
+		{ZPOOL_CONFIG_SCRUB_LAT_HISTO, offsetof(vdev_stat_ex_t,
+		    vsx_queue_histo[ZIO_PRIORITY_SCRUB])},
+	};
+
+	vsx = kmem_alloc(sizeof (*vsx), KM_SLEEP);
```

Comment:
This is a Solaris'ism but kmem_alloc(.., KM_SLEEP) is guaranteed to never fail.  Period.  We'll either get the memory or block here forever, so you can drop the error handling.